src/registrar/models/domain_invitation.py

```diff
@@ -0,0 +1,65 @@
+"""People are invited by email to administer domains."""
+
+from django.contrib.auth import get_user_model
+from django.db import models, IntegrityError
+
+from django_fsm import FSMField, transition  # type: ignore
+
+from .utility.time_stamped_model import TimeStampedModel
+from .user_domain_role import UserDomainRole
+
+
+class DomainInvitation(TimeStampedModel):
+    SENT = "sent"
+    RETRIEVED = "retrieved"
+
+    email = models.EmailField(
+        null=False,
+        blank=False,
+    )
+
+    domain = models.ForeignKey(
+        "registrar.Domain",
+        on_delete=models.CASCADE,  # delete domain, then get rid of invitations
+        null=False,
+        related_name="invitations",
+    )
+
+    status = FSMField(
+        choices=[
+            (SENT, SENT),
+            (RETRIEVED, RETRIEVED),
+        ],
+        default=SENT,
+        protected=True,  # can't alter state except through transition methods!
+    )
+
+    def __str__(self):
+        return f"Invitation for {self.email} on {self.domain} is {self.status}"
+
+    @transition(field="status", source=SENT, target=RETRIEVED)
+    def retrieve(self):
+        """When an invitation is retrieved, create the corresponding permission."""
+
+        # get a user with this email address
+        User = get_user_model()
+        try:
+            user = User.objects.get(email=self.email)
+        except User.DoesNotExist:
+            # should not happen because a matching user should exist before
+            # we retrieve this invitation
+            raise RuntimeError(
+                "Cannot find the user to retrieve this domain invitation."
+            )
```

Comment:
Since this is model code rather than view code, I think that we should raise exceptions here when assumptions are not met and be explicit in the function's documentation about the assumptions and the exceptions that could be raised.

The callers at the view level can decide how they want to handle those exceptions. @SSPJ does that sound like a reasonable way forward?

Reply:
Ideologically, I think errors should be handled close to where they occur.

Pragmatically, yes, documenting it (slight liking for [this style](https://android.googlesource.com/platform/external/google-styleguide/+/refs/tags/android-s-beta-2/pyguide.md#doc-function-raises) but no strong preference) is fine. The view you'll need to catch `RuntimeError` in is src/djangooidc/views.py at line 66. AFAIK, that's the only place it is called so far.